### PR TITLE
Linear function improved

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

Improved extraction of linear blocks in `CollectLinearFunctions` transpiler pass.

### Details and comments

The very first naive algorithm for collecting blocks of linear gates into linear functions simply looks for consecutive CX and SWAP gates after topologically sorting the nodes in DAG. This may miss opportunities to find larger blocks. 

Here is a specific example:
```
from qiskit.circuit.library import RealAmplitudes
from qiskit.transpiler.passes import (
    CollectLinearFunctions,
    LinearFunctionsSynthesis,
)
from qiskit.transpiler import PassManager

def test_real_amplitudes():
    ansatz = RealAmplitudes(4, reps=2, insert_barriers=False)
    circ1 = ansatz.decompose()
    print(f"Original        : {circ1.count_ops()}")
    print(circ1)
    circ2 = PassManager(CollectLinearFunctions()).run(circ1)
    print(f"Collect linear  : {circ2.count_ops()}")
    print(circ2)
    circ3 = PassManager(LinearFunctionsSynthesis()).run(circ2)
    print(f"Synthesize      : {circ3.count_ops()}")
    print(circ3)
```
This is the original circuit:
```
     ┌──────────┐                    ┌──────────┐                             »
q_0: ┤ Ry(θ[0]) ├──■────■─────────■──┤ Ry(θ[4]) ├─────────────────■────────■──»
     ├──────────┤┌─┴─┐  │         │  └──────────┘┌──────────┐   ┌─┴─┐      │  »
q_1: ┤ Ry(θ[1]) ├┤ X ├──┼────■────┼───────■──────┤ Ry(θ[5]) ├───┤ X ├──────┼──»
     ├──────────┤└───┘┌─┴─┐┌─┴─┐  │       │      └──────────┘┌──┴───┴───┐┌─┴─┐»
q_2: ┤ Ry(θ[2]) ├─────┤ X ├┤ X ├──┼───────┼───────────■──────┤ Ry(θ[6]) ├┤ X ├»
     ├──────────┤     └───┘└───┘┌─┴─┐   ┌─┴─┐       ┌─┴─┐    ├──────────┤└───┘»
q_3: ┤ Ry(θ[3]) ├───────────────┤ X ├───┤ X ├───────┤ X ├────┤ Ry(θ[7]) ├─────»
     └──────────┘               └───┘   └───┘       └───┘    └──────────┘     »
«               ┌──────────┐                         
«q_0: ───────■──┤ Ry(θ[8]) ├─────────────────────────
«            │  └──────────┘┌──────────┐             
«q_1: ──■────┼───────■──────┤ Ry(θ[9]) ├─────────────
«     ┌─┴─┐  │       │      └──────────┘┌───────────┐
«q_2: ┤ X ├──┼───────┼───────────■──────┤ Ry(θ[10]) ├
«     └───┘┌─┴─┐   ┌─┴─┐       ┌─┴─┐    ├───────────┤
«q_3: ─────┤ X ├───┤ X ├───────┤ X ├────┤ Ry(θ[11]) ├
«          └───┘   └───┘       └───┘    └───────────┘
```
and the circuit after `CollectLinearFunctions` pass:
```
     ┌──────────┐     ┌──────────────────┐     ┌──────────┐            »
q_0: ┤ Ry(θ[0]) ├──■──┤0                 ├──■──┤ Ry(θ[4]) ├────────────»
     ├──────────┤┌─┴─┐│                  │  │  └──────────┘┌──────────┐»
q_1: ┤ Ry(θ[1]) ├┤ X ├┤1 Linear_function ├──┼───────■──────┤ Ry(θ[5]) ├»
     ├──────────┤└───┘│                  │  │       │      └──────────┘»
q_2: ┤ Ry(θ[2]) ├─────┤2                 ├──┼───────┼──────────────────»
     ├──────────┤     └──────────────────┘┌─┴─┐   ┌─┴─┐                »
q_3: ┤ Ry(θ[3]) ├─────────────────────────┤ X ├───┤ X ├────────────────»
     └──────────┘                         └───┘   └───┘                »
«     ┌──────────────────┐            ┌──────────────────┐     ┌──────────┐»
«q_0: ┤0                 ├────────────┤0                 ├──■──┤ Ry(θ[8]) ├»
«     │                  │            │                  │  │  └──────────┘»
«q_1: ┤1                 ├────────────┤1 Linear_function ├──┼───────■──────»
«     │  Linear_function │┌──────────┐│                  │  │       │      »
«q_2: ┤2                 ├┤ Ry(θ[6]) ├┤2                 ├──┼───────┼──────»
«     │                  │├──────────┤└──────────────────┘┌─┴─┐   ┌─┴─┐    »
«q_3: ┤3                 ├┤ Ry(θ[7]) ├────────────────────┤ X ├───┤ X ├────»
«     └──────────────────┘└──────────┘                    └───┘   └───┘    »
«                              
«q_0: ─────────────────────────
«     ┌──────────┐             
«q_1: ┤ Ry(θ[9]) ├─────────────
«     └──────────┘┌───────────┐
«q_2: ─────■──────┤ Ry(θ[10]) ├
«        ┌─┴─┐    ├───────────┤
«q_3: ───┤ X ├────┤ Ry(θ[11]) ├
«        └───┘    └───────────┘
```
One can see that at the minimum it is possible to absorb the 3 CX gates around the first linear function. The reason why this does not happen is simple: the topological order is `Ry(θ[0]), Ry(θ[1]), CX(0,1), Ry(θ[2]), ...` and the very first `CX(0,1)` gets isolated.

I have improved the gate ordering. Starting with the topological order, we first move all possible non-linear gates upfront, then move all possible linear gates after these, then all non-linear gates, and so on. Currently, we allow moving a gate across another gate only when it's defined over a different set of qubits (and hence the two gates trivially commute). For the example above, this changes the transpiled circuit to:
```
     ┌──────────┐┌──────────────────┐┌──────────┐┌──────────────────┐»
q_0: ┤ Ry(θ[0]) ├┤0                 ├┤ Ry(θ[4]) ├┤0                 ├»
     ├──────────┤│                  │├──────────┤│                  │»
q_1: ┤ Ry(θ[1]) ├┤1                 ├┤ Ry(θ[5]) ├┤1                 ├»
     ├──────────┤│  Linear_function │├──────────┤│  Linear_function │»
q_2: ┤ Ry(θ[2]) ├┤2                 ├┤ Ry(θ[6]) ├┤2                 ├»
     ├──────────┤│                  │├──────────┤│                  │»
q_3: ┤ Ry(θ[3]) ├┤3                 ├┤ Ry(θ[7]) ├┤3                 ├»
     └──────────┘└──────────────────┘└──────────┘└──────────────────┘»
«      ┌──────────┐
«q_0: ─┤ Ry(θ[8]) ├
«      ├──────────┤
«q_1: ─┤ Ry(θ[9]) ├
«     ┌┴──────────┤
«q_2: ┤ Ry(θ[10]) ├
«     ├───────────┤
«q_3: ┤ Ry(θ[11]) ├
«     └───────────┘
```
Also note that the following `LinearFunctionsSynthesis` with the old `CollectLinearFunctions` produces the same number of CX gates as in the original circuit (12), while with the new  `CollectLinearFunctions` reduces the number of CX gates to 6. The code is reasonably fast on circuits with 100 qubits and above, and (in this `RealAmplitudes` example) the saving due to first collecting linear functions and then synthesizing linear function (using the Patel-Markov-Hayes algorithm) become significantly more pronounced as the number of qubits increases. I have also tried transpiling the original circuit both with `optimization_level=1` and `optimization_level=3`, however this did not improve the CX-count. 

One can probably take this even further by exploiting real commutativity between gates.

### CODE REVIEW

**Review by jakelishman — COMMENTED** [review]
This feels like quadratic complexity, by rescanning the nodes on each loop iteration.  At the very least, it adds a lot of overhead.

Can you achieve what you need by setting the `key` in `DAGCircuit.topological_nodes` to something that sorts linear gates ahead of all other gates, then defers to the normal mechanism?  That's probably something like
```python
def run(self, dag):
    ...
    def key(op):
        return (self._is_linear_function(op), op.sort_key)
    for node in dag.topological_op_nodes(key=key):
        ...
```

This is also trying to fix a bug, so it needs explicit tests to ensure that the fix works, and that we don't regress on the bug in the future.

**Comment by alexanderivrii** [comment]
@jakelishman Thanks! You are right, this is worst-case quadratic, since we can always scan till the end looking for an additional gate to add to the current block. I did try to put a few obvious optimizations: keeping track of the first unprocessed node (so that we don't need to repeatedly scan from the beginning of the circuit) and stopping the scan once every qubit is "contaminated". I did not know that one can put `key` in `topological_op_nodes`, this is definitely very useful. I want the sorting to first put all non-linear gates ahead of other gates, then all linear gates ahead of the remaining gates, then all non-linear gates, then all linear gates, and so on. Intuitively, this maximizes chances to find large linear blocks. I need to think more whether one can define a `key` for that...

**Comment by alexanderivrii** [comment]
I was not able to come up with a single `key` (as per @jakelishman 's suggestion), as the code currently alternates between collecting blocks of linear gates and blocks of non-linear gates. Instead I have added a simple cut-off heuristic that makes the algorithm worst-case linear and still works optimally on `RealAmplitudes` example above.

There was a short discussion about this in `qiskit-dev-synthesis` slack channel, and I believe that the overall consensus is that _in the long run_ we need a generic solution that can collect blocks of gates of various types and make use of commutativity analysis; in fact we already have an issue for this, see #5775. Because of this, I don't think it makes sense to invest a lot of time improving this specific `CollectLinearFunctions` pass.

On the other hand, _in the short run_, the slightly smarter `CollectLinearFunctions` from this PR is significantly more useful than the original naive implementation, as the `RealAmplitudes` example above demonstrates. I think it would be good to include it for `0.20`. 

Do we also want to add it into one of preset transpiler passes? (More precisely, the sequence `CollectLinearFunctions`, `LinearFunctionsToPermutations`, `LinearFunctionsSynthesis`). 

If there is a concern that the underlying Patel-Markov-Hayes algorithm is heuristic without any formal guarantees (as far as I know), I could add an option (similar to one @ajavadia had in the code for collecting swap gates that he shared in `qiskit-dev-synthesis` slack channel) that (when creating a linear function from a quantum circuit) would save the original definition, and that would use this original definition if it turns out to be better than the one produced by Patel-Markov-Hayes.

**Comment by georgios-ts** [comment]
Nice! You can combine your idea of alternating between collecting blocks of linear gates and blocks of non-linear gates into a modified version of [Kahn's algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm) for topological sorting with linear running time.

```python
from collections import deque
from qiskit.dagcircuit import DAGOpNode

def collect_linear_blocks(dag):
    blocks = []

    in_degree = dict()
    pending_linear_ops = deque()
    pending_non_linear_ops = deque()

    def process_node(node):
        for suc in dag.successors(node):
            if not isinstance(suc, DAGOpNode):
                continue
            in_degree[suc] -= 1
            if in_degree[suc] > 0:
                continue

            if suc.name in ['cx', 'swap']:
                pending_linear_ops.append(suc)
            else:
                pending_non_linear_ops.append(suc)


    for node in dag.op_nodes():
        deg = sum(1 for op in dag.predecessors(node) if isinstance(op, DAGOpNode))
        if deg > 0:
            in_degree[node] = deg
            continue
    
        if node.name in ['cx', 'swap']:
            pending_linear_ops.append(node)
        else:
            pending_non_linear_ops.append(node)
            

    while pending_non_linear_ops or pending_linear_ops:

        # first collect as many non linear gates as possible
        while pending_non_linear_ops:
            node = pending_non_linear_ops.popleft()
            process_node(node)
            
        # now collect as many linear gates as possible
        cur_block = []
        while pending_linear_ops:
            node = pending_linear_ops.popleft()
            process_node(node)
            cur_block.append(node)
        
        if cur_block:
            blocks.append(cur_block)

    return blocks
```

**Comment by alexanderivrii** [comment]
@georgios-ts , this is really elegant and really clever. Thanks! Taking your code as is.

File: qiskit/transpiler/passes/optimization/collect_linear_functions.py [hunk 1]
```diff
@@ -13,29 +13,70 @@
...
+            else:
+                pending_non_linear_ops.append(suc)
+
+    for node in dag.op_nodes():
+        deg = sum(1 for op in dag.predecessors(node) if isinstance(op, DAGOpNode))
+        if deg > 0:
+            in_degree[node] = deg
+            continue
+
+        if node.name in ["cx", "swap"]:
```
Comment (georgios-ts) [inline qiskit/transpiler/passes/optimization/collect_linear_functions.py:51]:
```suggestion
        if node.name in ["cx", "swap"] and node.condition is None:
```

File: qiskit/transpiler/passes/optimization/collect_linear_functions.py [hunk 2]
```diff
@@ -13,29 +13,70 @@
...
+            process_node(node)
+
+        # now collect as many linear gates as possible
+        cur_block = []
+        while pending_linear_ops:
+            node = pending_linear_ops.popleft()
+            process_node(node)
+            cur_block.append(node)
+
+        if cur_block:
```
Comment (georgios-ts) [inline qiskit/transpiler/passes/optimization/collect_linear_functions.py:70]: Again, you were checking that:

```suggestion
        # Collect only blocks comprising at least 2 gates.
        if len(cur_block) > 1:
```

**Comment by ajavadia** [comment]
> If there is a concern that the underlying Patel-Markov-Hayes algorithm is heuristic without any formal guarantees (as far as I know), I could add an option (similar to one @ajavadia had in the code for collecting swap gates that he shared in qiskit-dev-synthesis slack channel) that (when creating a linear function from a quantum circuit) would save the original definition, and that would use this original definition if it turns out to be better than the one produced by Patel-Markov-Hayes.

Yeah I think it's good to keep the original somewhere (not just for this, but for any blocks of gates that are consolidated), since synthesis algorithms sometime may not be able to improve on that.

**Comment by alexanderivrii** [comment]
@georgios-ts, many thanks!! The code is fixed now, and I have added tests to make sure that single linear gates are not turned into linear functions. I have added another simple improvement that splits the nodes into "connected components", for instance we would no longer create a linear function for `CX(0, 1); CX(2, 3)`, and again I have added appropriate tests. 

Based on @ajavadia's reply, I am also saving the original circuit definition, however I am not doing anything with this yet. We might later extend the `synthesize` method of `LinearFunction` to also include a metric on how different implementations should be compared (e.g., based on the number of 2-qubit gates, circuit-depth, etc.), and then we would be able to choose the _better_ synthesis method.

Again, all this is only a short-term solution that does not take gate commutativity into account, but I think it's slowly becoming more and more useful.

**Review by ShellyGarion — APPROVED** [review]

**Review by ajavadia — APPROVED** [review]

**Pull request merged**